File: 66.plus-one.java
/*
 * @lc app=leetcode id=66 lang=java
 *
 * [66] Plus One
 */

// @lc code=start
class Solution {
    public int[] plusOne(int[] digits) {

        int[] digitsP1 = new int[digits.length + 1];
        int carry = 1;

        for (int i=digits.length-1; i >= 0; i--) {
            int d = digits[i];

            int sum = d + carry;

            digitsP1[i+1] += sum % 10;

            carry = (sum > 9) ? 1 : 0;
        }

        if (carry > 0) digitsP1[0] += carry;

        return (digitsP1[0] > 0) ? digitsP1 : Arrays.copyOfRange(digitsP1, 1, digitsP1.length);
        
    }
}
// @lc code=end

File: 66.plus-one.c
/*
 * @lc app=leetcode id=66 lang=c
 *
 * [66] Plus One
 */

// @lc code=start


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* plusOne(int* digits, int digitsSize, int* returnSize){

    int i = digitsSize - 1;
    while (i >= 0) {
        if (digits[i] < 9) {
            digits[i]++;
            *returnSize = digitsSize;
            return digits;
        } else {
            digits[i] = 0;
            i--;
        }
    }

    *returnSize = digitsSize + 1;
    
    int* newDigits = (int*) calloc(*returnSize, sizeof(int)); // Dynamic array initialization using calloc

    // int* newDigits = (int*) malloc((*returnSize) * sizeof(int)); // Dynamic array initialization using malloc

    // int* newDigits = new int[*returnSize]; // C++ dynamic array initialization

    i = digitsSize;
    while (i >= 0) {
        newDigits[i--] = 0;
    }

    newDigits[0] = 1;

    return newDigits;
    



}
// @lc code=end

